pkg/drivers/jetstream/kv/kv.go

```diff
@@ -0,0 +1,342 @@
+	return &EncodedKV{bucket: bucket, keyCodec: k, valueCodec: v}
+}
+
+type WatcherWithCtx interface {
+	WatchWithCtx(ctx context.Context, keys string, opts ...nats.WatchOpt) nats.KeyWatcher
+}
+
+type KeyCodec interface {
+	Encode(key string) (string, error)
+	Decode(key string) (string, error)
+	EncodeRange(keys string) (string, error)
+}
+
+type ValueCodec interface {
+	Encode(src []byte, dst io.Writer) error
+	Decode(src io.Reader, dst io.Writer) error
+}
+
+type EncodedKV struct {
+	WatcherWithCtx
+	bucket     nats.KeyValue
+	keyCodec   KeyCodec
+	valueCodec ValueCodec
+}
+
+type watcher struct {
+	watcher    nats.KeyWatcher
+	keyCodec   KeyCodec
+	valueCodec ValueCodec
+	updates    chan nats.KeyValueEntry
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+func (w *watcher) Context() context.Context {
+	if w == nil {
+		return nil
+	}
+	return w.ctx
+}
+
+type entry struct {
+	keyCodec   KeyCodec
+	valueCodec ValueCodec
+	entry      nats.KeyValueEntry
+}
+
+func (e *entry) Key() string {
+	dk, err := e.keyCodec.Decode(e.entry.Key())
+	// should not happen
+	if err != nil {
+		// should not happen
+		logrus.Warnf("could not decode key %s: %v", e.entry.Key(), err)
+		return ""
+	}
+
+	return dk
+}
+
+func (e *entry) Bucket() string { return e.entry.Bucket() }
+func (e *entry) Value() []byte {
+	buf := new(bytes.Buffer)
+	if err := e.valueCodec.Decode(bytes.NewBuffer(e.entry.Value()), buf); err != nil {
+		// should not happen
+		logrus.Warnf("could not decode value for %s: %v", e.Key(), err)
+	}
+	return buf.Bytes()
+}
+func (e *entry) Revision() uint64           { return e.entry.Revision() }
+func (e *entry) Created() time.Time         { return e.entry.Created() }
+func (e *entry) Delta() uint64              { return e.entry.Delta() }
+func (e *entry) Operation() nats.KeyValueOp { return e.entry.Operation() }
+
+func (w *watcher) Updates() <-chan nats.KeyValueEntry { return w.updates }
+func (w *watcher) Stop() error {
+	if w.cancel != nil {
+		w.cancel()
+	}
+
+	return w.watcher.Stop()
+}
+
+func (e *EncodedKV) newWatcherWithCtx(ctx context.Context, w nats.KeyWatcher) nats.KeyWatcher {
+	watch := &watcher{
+		watcher:    w,
+		keyCodec:   e.keyCodec,
+		valueCodec: e.valueCodec,
+		updates:    make(chan nats.KeyValueEntry, 32)}
+
+	watch.ctx, watch.cancel = context.WithCancel(ctx)
+
+	go func() {
+		for {
+			select {
+			case ent := <-w.Updates():
+				if ent == nil {
+					watch.updates <- nil
+					continue
+				}
+
+				watch.updates <- &entry{
+					keyCodec:   e.keyCodec,
+					valueCodec: e.valueCodec,
+					entry:      ent,
+				}
+			case <-watch.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return watch
+}
+
+func (e *EncodedKV) newWatcher(w nats.KeyWatcher) nats.KeyWatcher {
+	watch := &watcher{
+		watcher:    w,
+		keyCodec:   e.keyCodec,
+		valueCodec: e.valueCodec,
+		updates:    make(chan nats.KeyValueEntry, 32)}
+
+	// waiting on https://github.com/nats-io/nats.go/pull/904 to refactor
```

Comment:
https://github.com/nats-io/nats.go/pull/904 has been merged; this can now be refactored